src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,186 @@
+        linkLast(value);
     }
 
     @Override
     public void add(T value, int index) {
+        checkIndexForAdd(index);
+        if (index == size()) {
+            linkLast(value);
+        } else {
+            linkBefore(value, findNode(index));
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T value : list) {
+            add(value);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> wanted = findNode(index);
+        return wanted.item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        Node<T> wanted = findNode(index);
+        T oldValue = wanted.item;
+        wanted.item = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        return unlink(findNode(index));
     }
 
     @Override
     public boolean remove(T object) {
+        if (object == null) {
+            for (Node<T> wanted = first; wanted != null; wanted = wanted.next) {
+                if (wanted.item == null) {
+                    unlink(wanted);
+                    return true;
+                }
+            }
+        } else {
+            for (Node<T> wanted = first; wanted != null; wanted = wanted.next) {
+                if (object.equals(wanted.item)) {
+                    unlink(wanted);
+                    return true;
+                }
+            }
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    @Override
+    public String toString() {
+        StringBuilder builder = new StringBuilder("[");
+        Node<T> current = first;
+        while (current != null) {
+            builder.append(current.item);
+            if (current.next != null) {
+                builder.append(", ");
+            } else {
+                builder.append("]");
+            }
+            current = current.next;
+        }
+        return builder.toString();
+    }
+
+    private void linkLast(T value) {
+        final Node<T> initLast = last;
+        final Node<T> newNode = new Node<>(initLast, value, null);
+        last = newNode;
+        if (initLast == null) {
+            first = newNode;
+        } else {
+            initLast.next = newNode;
+        }
+        size++;
+    }
+
+    private void linkBefore(T value, Node<T> after) {
+        final Node<T> before = after.prev;
+        final Node<T> newNode = new Node<>(before, value, after);
+        after.prev = newNode;
+        if (before == null) {
+            first = newNode;
+        } else {
+            before.next = newNode;
+        }
+        size++;
+    }
+
+    private T unlink(Node<T> found) {
```

Comment:
Be careful with linking Nodes. Make sure that you remove all the links when you remove a Node.